File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int answer;
int target_num;

void DFS(vector<int>& numbers, int num, int idx){
    if(idx == numbers.size()){
        if(num == target_num) answer++;
        return;
    }
    
    DFS(numbers, num + numbers[idx], idx + 1);
    DFS(numbers, num - numbers[idx], idx + 1);
}

int solution(vector<int> numbers, int target) {
    answer = 0;
    target_num = target;
    
    DFS(numbers, 0, 0);
    
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    
    int count;
    
    public int solution(int[] numbers, int target) {
        int answer = 0;
        
        count = 0;
        dfs(numbers, 0, 0, target);
        
        answer = count;
            
        return answer;
    }
    
    void dfs(int[] numbers, int idx, int value, int target) {
        if(numbers.length == idx) {
            if(value == target) {
                count++;
            }
            return;
        }
        
        dfs(numbers, idx + 1, value + numbers[idx], target);
        dfs(numbers, idx + 1, value - numbers[idx], target);
    }
}